In [2]:
import pandas as pd
df = pd.read_excel("Nifty50_with_last&first.xlsx", parse_dates=["Date"], index_col="Date")

In [2]:
df.head()

Open_last  Close_last  Low_last  High_last  Open_left  High_left  \
Date                                                                           
2010-01-04    5231.00      5235.0    5225.0    5238.00    5200.90    5238.45   
2010-01-05    5272.00      5274.0    5271.0    5288.00    5277.15    5288.35   
2010-01-06    5278.00      5281.0    5269.0    5287.00    5278.15    5310.85   
2010-01-07    5253.00      5256.6    5250.6    5275.85    5281.80    5302.55   
2010-01-08    5244.35      5247.0    5234.7    5255.60    5264.25    5276.75   

            Low_left  Close_left  Open_30  Close_30  Low_30  High_30  \
Date                                                                   
2010-01-04   5167.10      5232.2  5201.00   5220.00  5167.0  5224.00   
2010-01-05   5242.40      5277.9  5277.00   5278.00  5242.0  5286.00   
2010-01-06   5260.05      5281.8  5278.00   5287.00  5271.0  5311.00   
2010-01-07   5244.75      5263.1  5281.80   5270.15  5264.5  5302.55   
2010-01-08   5234.70      5244.7  5264.25   5259.65  5251.3  5275.10   

            Open_right  High_right  Low_right  Close_right       OPR  \
Date                                                                   
2010-01-04     5200.90     5238.45    5167.10       5232.2  0.473721   
2010-01-05     5277.15     5288.35    5242.40       5277.9  0.756257   
2010-01-06     5278.15     5310.85    5260.05       5281.8  0.356299   
2010-01-07     5281.80     5302.55    5244.75       5263.1  0.641003   
2010-01-08     5264.25     5276.75    5234.70       5244.7  0.702735   

                 CPR     Solid  True_Range_ratio  
Date                                              
2010-01-04  0.912404  0.438683          1.363671  
2010-01-05  0.772579  0.016322          0.870611  
2010-01-06  0.428150  0.071850          0.961793  
2010-01-07  0.317474  0.323529          1.098212  
2010-01-08  0.237812  0.464923          0.801762

In [3]:
df['TRR_Prev'] = df['True_Range_ratio'].shift(1)

In [4]:
df[['True_Range_ratio','TRR_Prev']].head()

True_Range_ratio  TRR_Prev
Date                                  
2010-01-04          1.363671       NaN
2010-01-05          0.870611  1.363671
2010-01-06          0.961793  0.870611
2010-01-07          1.098212  0.961793
2010-01-08          0.801762  1.098212

In [5]:
df['TRR_Next'] = df['True_Range_ratio'].shift(-1)
df['OPR_Next'] = df['OPR'].shift(-1)
df['CPR_Next'] = df['CPR'].shift(-1)
df['CPR_Prev'] = df['CPR'].shift(1)
df['OPR_Prev'] = df['OPR'].shift(1)
df['Prev_Close'] = df['Close_right'].shift(1)
df['Open-Low%'] = ((df['Open_right'] - df['Low_right'])/df['Prev_Close'])*100
df['Open-High%'] = ((df['High_right'] - df['Open_right'])/df['Prev_Close'])*100
df['Range~Prev_Close'] = ((df['High_right'] - df['Low_right'])/df['Prev_Close'])*100 

df['True_Range_ratio_30'] = ((df['High_30'] - df['Low_30'])/df['Prev_Close'])*100
df["OPR_30"] = (df['Open_30'] - df['Low_30'])/(df['High_30'] - df['Low_30'])
df["CPR_30"] = (df['Close_30'] - df['Low_30'])/(df['High_30'] - df['Low_30'])
df["Solid_30"] = (abs(df['Open_30'] - df['Close_30']))/(df['High_30'] - df['Low_30'])

df['Upper'] = ((df['High_right'] - df['High_30'])/df['Prev_Close'])*100
df['Lower'] = ((df['Low_30'] - df['Low_right'])/df['Prev_Close'])*100
df["Close~15min-High"] = ((df['Close_right'] - df['High_30'])/df['Prev_Close'])*100
df["Close~15min-Low"] = ((df['Low_30'] - df['Close_right'])/df['Prev_Close'])*100

## Range 

In [6]:
df['True_Range_ratio_last'] = ((df['High_last'] - df['Low_last'])/df['Prev_Close'])*100

In [7]:
df['True_Range_ratio_last'].describe()

count    2349.000000
mean        0.529721
std         0.327886
min         0.095820
25%         0.315996
50%         0.444390
75%         0.643337
max         3.542305
Name: True_Range_ratio_last, dtype: float64

In [8]:
df['True_Range_ratio_last'][ (df['OPR'] > 0.8) & (df['CPR'] < 0.2) ].describe()

count    333.000000
mean       0.630008
std        0.343097
min        0.149869
25%        0.376934
50%        0.548756
75%        0.763489
max        2.246013
Name: True_Range_ratio_last, dtype: float64

In [9]:
df['True_Range_ratio_last'][ (df['OPR'] < 0.2) & (df['CPR'] > 0.8) ].describe()

count    316.000000
mean       0.492883
std        0.268306
min        0.125253
25%        0.315063
50%        0.423191
75%        0.616241
max        2.591548
Name: True_Range_ratio_last, dtype: float64

In [10]:
df["OPR_last"] = (df['Open_last'] - df['Low_last'])/(df['High_last'] - df['Low_last'])
df["CPR_last"] = (df['Close_last'] - df['Low_last'])/(df['High_last'] - df['Low_last'])
df["Solid_last"] = (abs(df['Open_last'] - df['Close_last']))/(df['High_last'] - df['Low_last'])

In [21]:
df["OPR_last"].describe()

count    2350.000000
mean        0.498727
std         0.331891
min         0.000000
25%         0.181883
50%         0.500000
75%         0.814520
max         1.000000
Name: OPR_last, dtype: float64

In [26]:
df["OPR_last"][ (df['OPR'] > 0.8) & (df['CPR'] < 0.2) ].describe()

count    333.000000
mean       0.781379
std        0.215374
min        0.000000
25%        0.702381
50%        0.853392
75%        0.943311
max        1.000000
Name: OPR_last, dtype: float64

In [29]:
df["OPR_last"][ (df['OPR'] < 0.2) & (df['CPR'] > 0.8) ].describe()

count    316.000000
mean       0.249781
std        0.230053
min        0.000000
25%        0.062783
50%        0.182761
75%        0.359668
max        0.902685
Name: OPR_last, dtype: float64

#### This suggests that the momentum continues in the last hour of trending days

In [32]:
df["OPR_last"][ (((df['OPR_30'] > 0.3) & (df['CPR_30'] < 0.3)) | ((df['OPR_30'] > 0.8) & (df['CPR_30'] < 0.4))) ].describe()

count    783.000000
mean       0.526334
std        0.331878
min        0.000000
25%        0.209540
50%        0.560582
75%        0.843616
max        1.000000
Name: OPR_last, dtype: float64

In [33]:
df["OPR_last"][(((df['OPR_30'] < 0.7) & (df['CPR_30'] > 0.7)) | ((df['OPR_30'] < 0.2) & (df['CPR_30'] > 0.6)))].describe()

count    718.000000
mean       0.461464
std        0.328676
min        0.000000
25%        0.147726
50%        0.416410
75%        0.777151
max        1.000000
Name: OPR_last, dtype: float64

#### This suggests that there is no relation between last hour and first 15min candle

In [22]:
df["CPR_last"].describe()

count    2350.000000
mean        0.499154
std         0.273702
min         0.000000
25%         0.259335
50%         0.503819
75%         0.736570
max         1.000000
Name: CPR_last, dtype: float64

In [27]:
df["CPR_last"][ (df['OPR'] > 0.8) & (df['CPR'] < 0.2) ].describe()

count    333.000000
mean       0.288172
std        0.194166
min        0.000000
25%        0.145363
50%        0.254417
75%        0.396319
max        0.937025
Name: CPR_last, dtype: float64

In [28]:
df["CPR_last"][ (df['OPR'] < 0.2) & (df['CPR'] > 0.8) ].describe()

count    316.000000
mean       0.662498
std        0.225512
min        0.000000
25%        0.532690
50%        0.719737
75%        0.831902
max        0.987437
Name: CPR_last, dtype: float64

In [24]:
df["Solid_last"].describe()

count    2350.000000
mean        0.459423
std         0.254786
min         0.000000
25%         0.245438
50%         0.467748
75%         0.671022
max         0.994614
Name: Solid_last, dtype: float64

#### This suggests that overall on an average, there is no edge towards the end of day

In [25]:
df["Solid_last"][ (df['OPR'] > 0.8) & (df['CPR'] < 0.2) ].describe()

count    333.000000
mean       0.535344
std        0.248814
min        0.006073
25%        0.343214
50%        0.573050
75%        0.719677
max        0.981923
Name: Solid_last, dtype: float64

In [30]:
df["Solid_last"][ (df['OPR'] < 0.2) & (df['CPR'] > 0.8) ].describe()

count    316.000000
mean       0.483170
std        0.264659
min        0.000000
25%        0.273324
50%        0.505969
75%        0.712284
max        0.966734
Name: Solid_last, dtype: float64

#### CPR_last and OPR_last does suggest continued momentum wrt the direction of the day, in case of trending days



In [12]:
df["Solid_last"][ (df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.447435
std        0.256474
min        0.000000
25%        0.228571
50%        0.449102
75%        0.665299
max        0.970909
Name: Solid_last, dtype: float64

In [13]:
df["Solid_last"][ (df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.459850
std        0.252777
min        0.001980
25%        0.248217
50%        0.471318
75%        0.662939
max        0.981923
Name: Solid_last, dtype: float64

#### The above stats clearly suggests that things tend to turn dicey in the last hour of the day

In [14]:
df['OPR_last'][ (df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.596869
std        0.316005
min        0.000000
25%        0.321321
50%        0.682646
75%        0.877419
max        1.000000
Name: OPR_last, dtype: float64

In [15]:
df['OPR_last'][ (df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.380992
std        0.299749
min        0.000000
25%        0.116124
50%        0.317757
75%        0.634723
max        0.994495
Name: OPR_last, dtype: float64

In [16]:
df['CPR_last'][ (df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.429413
std        0.264529
min        0.000000
25%        0.209524
50%        0.389231
75%        0.655769
max        1.000000
Name: CPR_last, dtype: float64

In [17]:
df['CPR_last'][ (df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.561404
std        0.268708
min        0.000000
25%        0.343337
50%        0.608896
75%        0.788697
max        1.000000
Name: CPR_last, dtype: float64

In [18]:
df['True_Range_ratio_last'][ (df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.486170
std        0.275607
min        0.125253
25%        0.295125
50%        0.411589
75%        0.586037
max        2.591548
Name: True_Range_ratio_last, dtype: float64

In [19]:
df['True_Range_ratio_last'][ (df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.556897
std        0.339906
min        0.118993
25%        0.337009
50%        0.471282
75%        0.667408
max        3.542305
Name: True_Range_ratio_last, dtype: float64

In [20]:
df['Upper_last'] = ((df['High_last'] - df['Open_last'])/df['Prev_Close'])*100
df['Lower_last'] = ((df['Open_last'] - df['Low_last'])/df['Prev_Close'])*100

In [21]:
df['Upper_last'].describe()

count    2349.000000
mean        0.254404
std         0.253601
min         0.000000
25%         0.079615
50%         0.182914
75%         0.344340
max         2.212123
Name: Upper_last, dtype: float64

In [22]:
df['Upper_last'][ (df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.310171
std        0.260971
min        0.005080
25%        0.119928
50%        0.237500
75%        0.405656
max        1.943510
Name: Upper_last, dtype: float64

In [36]:
df['Upper_last'][ (df['OPR'] < 0.25) & (df['Upper_last'] > 0.3)].describe()

count    240.000000
mean       0.555358
std        0.247808
min        0.300054
25%        0.372981
50%        0.477009
75%        0.641414
max        1.943510
Name: Upper_last, dtype: float64

In [37]:
df['Upper_last'][ (df['OPR'] < 0.25) & (df['Upper_last'] > 0.6)].describe()

count    77.000000
mean      0.843881
std       0.233205
min       0.602532
25%       0.661924
50%       0.832574
75%       0.937326
max       1.943510
Name: Upper_last, dtype: float64

In [26]:
df['Lower_last'][ (df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.175998
std        0.200141
min        0.000000
25%        0.044925
50%        0.114127
75%        0.231621
max        1.550077
Name: Lower_last, dtype: float64

In [23]:
df['Lower_last'].describe()

count    2349.000000
mean        0.275318
std         0.304862
min         0.000000
25%         0.073181
50%         0.186473
75%         0.371783
max         2.825641
Name: Lower_last, dtype: float64

In [24]:
df['Lower_last'][ (df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.356219
std        0.344194
min        0.000000
25%        0.117096
50%        0.261509
75%        0.485928
max        2.825641
Name: Lower_last, dtype: float64

In [38]:
df['Lower_last'][ (df['OPR'] > 0.75) & (df['Lower_last'] > 0.3)].describe()

count    386.000000
mean       0.616865
std        0.349970
min        0.300442
25%        0.388371
50%        0.500946
75%        0.719638
max        2.825641
Name: Lower_last, dtype: float64

In [39]:
df['Lower_last'][ (df['OPR'] > 0.75) & (df['Lower_last'] > 0.6)].describe()

count    142.000000
mean       0.951441
std        0.380349
min        0.601369
25%        0.696594
50%        0.835981
75%        1.111268
max        2.825641
Name: Lower_last, dtype: float64

In [25]:
df['Upper_last'][ (df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.200678
std        0.200838
min        0.000000
25%        0.061807
50%        0.141097
75%        0.276254
max        1.326061
Name: Upper_last, dtype: float64

#### If u have general idea about the OPR, then it is fairly evident that the index will continue to move in the direction of trend 

In [29]:
df["Close~High"] = ((df['High_last'] - df['Close_last'])/df['Prev_Close'])*100
df["Close~Low"] = ((df['Close_last'] - df['Low_last'])/df['Prev_Close'])*100

In [30]:
df["Close~Low"][(df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.226194
std        0.204561
min        0.000000
25%        0.090500
50%        0.165061
75%        0.289578
max        1.434349
Name: Close~Low, dtype: float64

In [31]:
df["Close~Low"][(df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.287423
std        0.249097
min        0.000000
25%        0.108522
50%        0.223963
75%        0.378912
max        2.008858
Name: Close~Low, dtype: float64

In [32]:
df["Close~High"][(df['OPR'] > 0.75)].describe()

count    837.000000
mean       0.330703
std        0.302634
min        0.000000
25%        0.132380
50%        0.243750
75%        0.437003
max        2.982908
Name: Close~High, dtype: float64

In [33]:
df["Close~High"][(df['OPR'] < 0.25)].describe()

count    603.000000
mean       0.198746
std        0.177326
min        0.000000
25%        0.087059
50%        0.146060
75%        0.249527
max        1.266945
Name: Close~High, dtype: float64

#### So wrt close, there is 0.1% edge

### Conclusion :
#### On bullish days, on an average it travels 0.3%in the direction of the trend. And, 25% of the times it moves more than 0.4%.
#### On bearish days, on an average it travels 0.35%in the direction of the trend. And, 25% of the times it moves more than 0.48%.

## Plan of Action :

#### According to current market scenario, CMP = 17000
#### Therefore, 0.3% = 51 and 0.4% = 68 and 0.5% = 85pts

#### If we buy options 50 or 100 points away from the strike then,

#### Statistics suggest that on the bullish side if the average premium is less than 16 and on the bearish side if it is less than 20 for 50 point wide strike. Then it makes perfect sense to buy options 50points wide.

#### Statistics suggest that on the bullish side if the average premium is less than 5.3 and on the bearish side if it is less than 10.7 for 100 point wide strike. Then it makes perfect sense to buy options 100 points wide.